src/process/process_set.cu

```diff
@@ -0,0 +1,135 @@
+#include <iostream>
+
+namespace micm {
+    namespace cuda {
+      // flipped memory layout
+    __global__ void AddForcingTerms_kernel(
+        double* rate_constants, 
+        double* state_variables, 
+        double* forcing, 
+        int ngrids,
+        int nrxns, 
+        int nspecs, 
+        size_t* number_of_reactants_, 
+        size_t* reactant_ids_, 
+        size_t* number_of_products_, 
+        size_t* product_ids_, 
+        double* yields_)
+    {
+      //define thread index 
+      size_t tid = blockIdx.x * blockDim.x + threadIdx.x; 
+      size_t react_id_offset, prod_id_offset, yield_offset;
+
+      if (tid < ngrids){
+         react_id_offset = 0;
+	 prod_id_offset = 0;
+	 yield_offset = 0;
+         for (std::size_t i_rxn = 0; i_rxn < nrxns; ++i_rxn)
+         {
+           double rate = rate_constants[i_rxn*ngrids+tid];
+           size_t index;
+
+           for (std::size_t i_react = 0; i_react < number_of_reactants_[i_rxn]; ++i_react)
+             rate *= state_variables[reactant_ids_[react_id_offset+i_react]*ngrids+tid];
+           for (std::size_t i_react = 0; i_react < number_of_reactants_[i_rxn]; ++i_react){
+             forcing[reactant_ids_[react_id_offset+i_react]*ngrids+tid] -= rate;
+	   }
+           for (std::size_t i_prod = 0; i_prod < number_of_products_[i_rxn]; ++i_prod){
+             index = product_ids_[prod_id_offset+i_prod]*ngrids+tid;
+             forcing[index] += yields_[yield_offset+i_prod] * rate;
+           }
+           react_id_offset += number_of_reactants_[i_rxn];
+           prod_id_offset += number_of_products_[i_rxn];
+           yield_offset += number_of_products_[i_rxn];
+         } // for loop over number of reactions
+      }    // if check for valid CUDA threads
+    }      // end of AddForcingTerms_kernel
+
+    void AddForcingTerms_kernelSetup(
+        const double* rate_constants_data,
+        const double* state_variables_data,
+        double* forcing_data,
+        int ngrids,
+        int nrxns,
+        int nspecs,
+        const size_t* number_of_reactants,
+        int number_of_reactants_size,
+        const size_t* reactant_ids, 
+        int reactant_ids_size,
+        const size_t* number_of_products, 
+        int number_of_products_size,
+        const size_t* product_ids,
+        int product_ids_size,
+        const double* yields,
+        int yields_size){
```

Comment:
I thought @K20shores implemented a version that calls an additional function to get those constants. Thus CPU and GPU code could share the same interface for the `AddForcingTerms_*` function. Is it not adopted later for some reasons?

Reply:
I think the interface for AddForcingTerms function was CudaAddForcingTerms() which calls AddForcingTerms_kernelSetup() internally. We still needs to pass in those constants as arguments. Correct me if I am wrong? @K20shores @mattldawson

Reply:
Gotcha! Thanks for the clarification.